extensions/resume.md

```diff
@@ -0,0 +1,306 @@
+
+#### `BRB` Message
+This message, sent from a server to a client, indicates that a `BRB` request has been successful.
+
+    BRB <timeout>
+
+`<timeout>` is an integer that represents how many seconds the client has to reconnect before their session is closed and they will no longer be able to resume. If the server cannot give an accurate duration, it should either approximate the length of time or give a zero `0` (representing that the server cannot estimate how long until the session is terminated).
+
+After this message is returned, the client's connection to the server is immediately terminated.
+
+
+## Capability Negotiation
+The first step in successfully resuming a connection is for the 'old client' to negotiate the `draft/resume-0.4` capability and receive a resume token. This process is illustrated here:
+
+    C: CAP LS
+    S: CAP * LS :multi-prefix draft/resume-0.4
+    C: CAP REQ :draft/resume-0.4
+    S: CAP * ACK :draft/resume-0.4
+    S: RESUME TOKEN JKbAypzFiovffzuD8VEfcs6bOLrXsSenxsyZNt87Bb0YaQlyOzuQciz6cB2R
+
+Considerations around tokens and the process for generating them is described below in the [Resume Token](#resume-token) section.
+
+
+## Resuming A Connection
+When a client detects that it has become disconnected from a server, it SHOULD try to resume before it breaks the existing connection.
+
+Upon establishing the new connection, the client begins capability negotiation, negotiates all mutually-supported capabilities, and MUST confirm that the `draft/resume-0.4` capability exists. If this capability does not exist, the client continues connection registration without attempting to resume. If this capability does exist, the client sends the `RESUME` command and MUST wait for either a `RESUME` or a `FAIL RESUME` message from the server before continuing registration. It should be noted that the client MUST NOT perform SASL authentication if the `draft/resume-0.4` capability exists and they wish to resume their session, as completing SASL auth will end connection registration and abort the resumption attempt.
+
+If the token provided by the new client is validated by the server, the old client completed connection registration with the server, and both the old and new clients use TLS, then the attempt SHOULD be successful. If the attempt is successful, the server MUST send the client a `RESUME` message and complete connection registration immediately (at which time the state will begin to replay as described below). If the attempt is unsuccessful (for example, if either session is not using a secure connection), the server MUST send a `FAIL RESUME` message, and then allow the client to continue connection registration.
+
+If the client receives a `RESUME SUCCESS` message, connection registration will complete immediately and state will begin to replay as described below. If the client receives a `FAIL RESUME`, the client MUST continue registration as though they are joining the server normally.
```

Comment:
This disagrees with the description of s2c commands. `RESUME <nick>` is described above as being a success.

Reply:
I would be in favor of bringing back `RESUME SUCCESS <nick>` (that's what I had in my `draft/resume-0.3` implementation.

So the server-to-client message would be:

1. `RESUME TOKEN`
1. `RESUME SUCCESS`
1. `RESUMED`
1. `FAIL RESUME`, `WARN RESUME`, and friends (according to the standard replies extension)